ci: Add GitHub Actions workflow

jtojnar opened the pull request:
Depends on #20.

This will run tests on all supported PHP version 5.6 through 8.0 and check the coding style.

I have decided to use [Nix](https://nixos.org/) to install different PHP versions since it is a tool I am most familiar with and allows using the same environment locally as on CI. (Think Docker but much more flexible.)

Nix provides `nix-shell` command which will enter a shell with all the dependencies installed into environment. And if you use it with [`direnv`](https://github.com/direnv/direnv), you will be put into the environment automatically whenever you `cd` into the project directory in a terminal emulator.

You can also switch the used PHP version by changing `matrix.phpPackage` value in the `flake.nix` file and re-running `nix-shell` (or `touch shell.nix` if you use `direnv`).

For more information, check out fossar/selfoss@e0a0ee2/docs/content/docs/development/using-nix.md

jtojnar committed:
ci: Add Nix files

This will allow people to easily enter a shell with dependencies like PHP and composer pre-installed.

Just install *Nix package manager* and run `nix-shell`.
You can also get *direnv* to be placed in the environment automatically
whenever you `cd` into the project directory in a terminal emulator.

You can also switch the used PHP version by changing `matrix.phpPackage` value
in the `flake.nix` file and re-running `nix-shell`.

For more information, check out https://github.com/fossar/selfoss/blob/e0a0ee2a88fc1be8563a3eb361c7b1142b239232/docs/content/docs/development/using-nix.md

jtojnar committed:
ci: Add GitHub Actions workflow

This will run tests on all supported PHP version 5.6 through 8.0 and check the coding style.

jtojnar committed:
composer: Add a test script

This will allow running tests using `composer test` instead of the hard to remember `sh test/run.sh`.

smottt commented:
I like this, will check once we get the rest sorted out. I have a local set up of some docker containers for multiple PHP versions which also include Xdebug etc and I can configure PHPStorm accordingly. Is it possible to have interpreter & debugger configured correctly with this?

jtojnar commented:
You can enable the xdebug extension as follows:

```nix
--- a/flake.nix
+++ b/flake.nix
@@ -26,7 +26,9 @@
         pkgs = nixpkgs.legacyPackages.${system};
 
         # Create a PHP package from the selected PHP package.
-        php = phps.packages.${system}.${matrix.phpPackage};
+        php = phps.packages.${system}.${matrix.phpPackage}.withExtensions ({ enabled, all }: with all; enabled ++ [
+          xdebug
+        ]);
       in {
         # Expose shell environment for development.
         devShell.${system} = pkgs.mkShell {

```

PHP in the shell then has the extension loaded:

```ShellSession
$ php -m | grep xdebug
xdebug
```

Not sure what else needs to be done to set it up for PHPStorm.

jtojnar commented:
@smottt I cherry-picked changes from #19 and #20 and now the tests pass everywhere:

Had to do some extra processing since PHPUnit 9 does not like `assertInternalType` ([removed](https://github.com/sebastianbergmann/phpunit/issues/3368)) or `@expectedException` annotation (not sure why).

smottt commented:
`@expectedException` has been deprecated years ago, I guess it was also removed at some point.

I think I'll modify the formula for BMP size based on your comment (so it's easier to add support for other than 24-bit if we want that at some point in the future) and merge #19, then #20 and let's see what's left here afterwards.

jtojnar commented:
Not sure if adding more bit depths is worth it GD supports BMP since PHP 7.2.

smottt commented:
I forgot about that. Seems that TGA is also supported from 7.4 on, so could introduce native support for those two without custom mappers then at some point. Thanks for the tip!

jtojnar committed:
ci: Use php-cs-fixer from composer

This tool works with PHP 5.6+ so there should hopefully be no issues installing it through composer.

smottt commented on .github/workflows/main.yaml:
```diff
@@ -0,0 +1,64 @@
+      - name: Set up Nix cache
+        uses: cachix/cachix-action@v8
+        with:
+          name: fossar
```
Is this correct?

jtojnar replied:
The way Nix works, it takes build instructions for packages and builds them. It can also substitute the build result from a cache to reduce work. But since these are custom packages, they are not built by the official build servers and we need to use our own cache. I added a comment with a link to the repository that builds it.

smottt commented on composer.json:
```diff
@@ -18,10 +18,17 @@
-    "config": {
-        "bin-dir": "bin"
+    "scripts": {
+        "cs": "php-cs-fixer fix --verbose --dry-run --diff",
+        "fix": "php-cs-fixer fix --verbose --diff",
```
Could these two be renamed to be more clear?

Like `cs-check` and `cs-fix` perhaps?

smottt commented on flake.lock:
```diff
@@ -0,0 +1,78 @@
+{
```
Usually lock files are not committed for libraries. Not sure if this should be here or not, though?

jtojnar replied:
This is lockfile for the CI, not the library. I find it more reliable to have CI state locked and it will also allow you to use the exact same version of dependencies locally in `nix-shell`.

jtojnar committed:
cs: Fix php-cs-fixer complaint
